File: app/src/main/java/com/awscherb/cardkeeper/ui/scan/ScanFragment.java
package com.awscherb.cardkeeper.ui.scan;

import android.Manifest;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.PermissionChecker;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.awscherb.cardkeeper.R;
import com.awscherb.cardkeeper.ui.base.BaseFragment;
import com.google.zxing.ResultPoint;
import com.journeyapps.barcodescanner.BarcodeCallback;
import com.journeyapps.barcodescanner.BarcodeResult;
import com.journeyapps.barcodescanner.CompoundBarcodeView;

import java.util.List;

import static android.app.Activity.RESULT_OK;

public class ScanFragment extends BaseFragment {

    private static final String TAG = ScanFragment.class.getName();

    public static final String EXTRA_BARCODE_FORMAT = TAG + ".extra_barcode_format";
    public static final String EXTRA_BARCODE_TEXT = TAG + ".extra_barcode_text";

    private static final int REQUEST_CAMERA = 16;

    private CompoundBarcodeView barcodeView;

    //================================================================================
    // Lifecycle methods
    //================================================================================

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_scan, container, false);
        if (v == null) {
            return null;
        }

        barcodeView = (CompoundBarcodeView) v.findViewById(R.id.fragment_scan_scanner);

        // Check permissions
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (getActivity().checkSelfPermission(Manifest.permission.CAMERA) == PermissionChecker.PERMISSION_GRANTED) {
                barcodeView.resume();
            } else {
                requestPermissions(new String[]{Manifest.permission.CAMERA},
                        REQUEST_CAMERA);
            }
        }

        // Setup scanner
        setCallback();

        return v;
    }

    @Override
    public void onResume() {
        super.onResume();
        barcodeView.resume();
    }

    @Override
    public void onPause() {
        super.onPause();
        barcodeView.pause();
    }

    // ========================================================================
    // Helper methods
    // ========================================================================

    private void setCallback() {
        BarcodeCallback callback = new BarcodeCallback() {
            @Override
            public void barcodeResult(BarcodeResult result) {

                String text = "";
                if (result.getText() != null) {
                    text = result.getText();
                }

                Intent data = new Intent();
                data.putExtra(EXTRA_BARCODE_FORMAT, result.getBarcodeFormat());
                data.putExtra(EXTRA_BARCODE_TEXT, text);
                getActivity().setResult(RESULT_OK, data);
                getActivity().finish();
            }

            @Override
            public void possibleResultPoints(List<ResultPoint> resultPoints) {
            }
        };

        barcodeView.decodeContinuous(callback);
    }
}

File: app/src/main/java/com/awscherb/cardkeeper/ui/scan/ScanFragment.kt
package com.awscherb.cardkeeper.ui.scan

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.awscherb.cardkeeper.R
import com.awscherb.cardkeeper.ui.base.BaseFragment
import com.google.zxing.ResultPoint
import com.journeyapps.barcodescanner.BarcodeCallback
import com.journeyapps.barcodescanner.BarcodeResult
import com.journeyapps.barcodescanner.CompoundBarcodeView
import dagger.android.support.AndroidSupportInjection
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject

class ScanFragment : BaseFragment() {

    private val viewModel by viewModels<ScanViewModel> { factory }

    @Inject
    lateinit var factory: ScanViewModelFactory

    private lateinit var scannerView: CompoundBarcodeView

    private val found = AtomicBoolean(false)

    //================================================================================
    // Lifecycle methods
    //================================================================================

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = inflater.inflate(R.layout.fragment_scan, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        AndroidSupportInjection.inject(this)

        scannerView = view.findViewById(R.id.fragment_scan_scanner)

        // Check permissions
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (activity?.checkSelfPermission(Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
                scannerView.resume()
            } else {
                requestPermissions(
                    arrayOf(Manifest.permission.CAMERA),
                    REQUEST_CAMERA
                )
            }
        }

        // Setup scanner
        setCallback()

        viewModel.createResult.observe(viewLifecycleOwner) {
            onCodeAdded()
        }
    }

    override fun onResume() {
        super.onResume()
        scannerView.resume()
    }

    override fun onPause() {
        super.onPause()
        scannerView.pause()
    }

    private fun onCodeAdded() {
        findNavController().navigate(
            ScanFragmentDirections.actionScanFragmentToCardsFragment()
        )
    }

    // ========================================================================
    // Helper methods
    // ========================================================================

    private fun setCallback() {
        val callback = object : BarcodeCallback {
            override fun barcodeResult(result: BarcodeResult) {
                if (found.compareAndSet(false, true)) {
                    val input = EditText(activity).apply {
                        setHint(R.string.dialog_card_name_hint)
                        inputType = InputType.TYPE_TEXT_FLAG_CAP_WORDS
                    }

                    AlertDialog.Builder(activity!!)
                        .setTitle(R.string.app_name)
                        .setView(input)
                        .setOnDismissListener { found.set(false) }
                        .setPositiveButton(R.string.action_add) { _, _ ->
                            viewModel.createData.postValue(
                                CreateCodeData(
                                    format = result.barcodeFormat,
                                    text = result.text,
                                    title = input.text.toString()
                                )
                            )
                        }
                        .setNegativeButton(R.string.action_cancel) { dialog, _ -> dialog.dismiss() }
                        .show()
                }
            }

            override fun possibleResultPoints(resultPoints: List<ResultPoint>) {}
        }

        scannerView.decodeContinuous(callback)
    }

    companion object {
        private const val REQUEST_CAMERA = 16
    }
}
